dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+                }
+                else
+                {
+                    channelRequestResult = channelRequestException.code();
+                    status = PushNotificationChannelStatus::CompletedFailure;
+                    break;
+                }
+            }
+
+            co_await winrt::resume_after(std::chrono::seconds(backOffTimeInSeconds));
+        }
+
+
+        if (status == PushNotificationChannelStatus::CompletedSuccess)
+        {
+            // Returns a com_ptr to the implementation type
+            auto pushChannel =
+                winrt::make_self<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationChannel>(pushChannelReceived);
+
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                *(pushChannel.get()),
+                channelRequestResult,
+                status);
+        }
+        else if (status == PushNotificationChannelStatus::CompletedFailure)
+        {
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                nullptr,
+                channelRequestResult,
+                status);
+        }
+
+        co_return channelResult;
+    }
+
+    PushNotificationRegistrationToken PushNotificationManager::RegisterActivator(PushNotificationActivationInfo const& details)
+    {
+        winrt::guid taskClsid = details.TaskClsid();
+        DWORD cookie = 0;
+        BackgroundTaskRegistration registeredTask = nullptr;
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::PushTrigger))
+        {
+            bool taskRegistered = false;
+            for (auto task : BackgroundTaskRegistration::AllTasks())
+            {
+                if (task.Value().Name() == backgroundTaskName)
+                {
+                    taskRegistered = true;
+                    break;
+                }
+            }
+
+            if (!taskRegistered)
+            {
+                BackgroundTaskBuilder builder;
+                builder.Name(backgroundTaskName);
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                if (taskClsid != winrt::guid())
+                {
+                    auto builder5 = builder.try_as<winrt::IBackgroundTaskBuilder5>();
+                    if (IsPackagedProcess() && builder5)
+                    {
+                        builder5.SetTaskEntryPointClsid(taskClsid);
+                        winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                        for (auto condition : conditions)
+                        {
+                            builder.AddCondition(condition);
+                        }
+                    }
+                }
+
+                registeredTask = builder.Register();
+            }
+        }
```

Comment:
What if the task is already registered, but with some other `taskClsid`? We will return success even though the caller's `taskClsid` will never be called. (Instead, we're going to call some other `taskClsid`.)

Reply:
IMHO, we should prevent the app from calling Register with a different CLSID. 
@pmpurifoy If a BackgroundTask already exists for this app, can we check the registered CLSID of that task and compare it with the passed value? IF they are not the same, I believe we should throw here.

Reply:
What error should be thrown for this scenario?

Reply:
We are already throwing invalid argument:

`throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");`

If this is not correct, please reopen the comment and make a suggestion.